### PR TITLE
Add custom headers to smart http requests

### PR DESCRIPTION
**Pull request by spraints**
This integrates https://github.com/libgit2/libgit2/pull/3411 into rugged.

### CODE REVIEW

**Comment by arthurschreiber** [comment]
@spraints Cool stuff!

As-is, this is going to leak memory on every call. You need call `git_strarray_free()` after you're done with the `git_strarray` struct.

You need to do something like this:

```c
static VALUE rb_git_remote_extra_http_headers(VALUE self)
{
	git_remote *remote;
	git_strarray http_headers;
	VALUE rb_result;


	Data_Get_Struct(self, git_remote, remote);

	rugged_exception_check(
		git_remote_extra_http_headers(remote, &http_headers)
	);

	rb_result = rugged_strarray_to_rb_ary(&http_headers);
	git_strarray_free(&http_headers);

	return rb_result;
}
```

**Comment by spraints** [comment]
> As-is, this is going to leak memory on every call. You need call git_strarray_free() after you're done with the git_strarray struct.

Ah, cool. Good catch. libgit2/libgit2#3411 needs some more iteration. When that's more settled, I'll come back to this.

File: ext/rugged/rugged_remote.c [hunk 1]
```diff
@@ -561,6 +561,17 @@ static VALUE rb_git_remote_fetch(int argc, VALUE *argv, VALUE self)
 	return rb_result;
 }
 
+void init_push_headers(VALUE rb_options, git_strarray *custom_headers)
+{
+	VALUE rb_headers;
```
Comment (arthurschreiber) [inline ext/rugged/rugged_remote.c]: You can move this declaration inside the `if` block.

File: ext/rugged/rugged_remote.c [hunk 2]
```diff
@@ -561,6 +561,17 @@ static VALUE rb_git_remote_fetch(int argc, VALUE *argv, VALUE self)
 	return rb_result;
 }
 
+void init_push_headers(VALUE rb_options, git_strarray *custom_headers)
```
Comment (arthurschreiber) [inline ext/rugged/rugged_remote.c]: This can be `static`, I think.

File: ext/rugged/rugged_remote.c (hunk 1 above)
Comment (spraints) [inline ext/rugged/rugged_remote.c]: > You can move this declaration inside the if block.

1324d66

File: ext/rugged/rugged_remote.c (hunk 2 above)
Comment (spraints) [inline ext/rugged/rugged_remote.c]: TIL `static`. 4c741f6

**Comment by arthurschreiber** [comment]
Looking great. :+1:

**Comment by arthurschreiber** [comment]
@spraints Can you bump libgit2 and squash the other commits so this can be merged?

**Comment by spraints** [comment]
> Can you bump libgit2 and squash the other commits so this can be merged?

Yep! fde2448 is squashed with an updated libgit2. :grinning:

Looking at the diff now, it looks like I could add support for custom fetch headers too. IIRC, I added that to https://github.com/libgit2/libgit2/pull/3411, but it isn't here. Would you like me to do that?

**Comment by spraints** [comment]
> Looking at the diff now, it looks like I could add support for custom fetch headers too. IIRC, I added that to libgit2/libgit2#3411, but it isn't here. Would you like me to do that?

2f1f02d adds custom headers in more places. If it looks good, I can squash again.

**Comment by arthurschreiber** [comment]
Oh, good catch with freeing of the strarray objects. Everything LGTM!

**Comment by spraints** [comment]
> Oh, good catch with freeing of the strarray objects. Everything LGTM!

:metal: It's now squashed down to one commit. :smile_cat:

**Comment by arthurschreiber** [comment]
PEEEEEEEEEEEEEEERFEEEEECT! :heart:

**Pull request merged**